hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationLog.java

```diff
@@ -171,36 +173,37 @@ public void fromJson(JsonObject json) {
     }
 
     static final class Invocation implements JsonSerializable {
+        // values are just set once, but cannot be final due to JsonSerializable
         int id;
         long startedAt;
-        long durationNanos;
-        volatile long lastAccessNanos;
 
-        public long getDurationNanos() {
-            return durationNanos;
+        // only accessed by single SlowOperationDetectorThread, no need for synchronization
+        long lastAccessNanos;
+
+        // this field will be changed by single SlowOperationDetectorThread only (they can be read by multiple threads)
+        volatile int durationMs;
+
+        public long getDurationMs() {
+            return durationMs;
         }
 
         @Override
         public String toString() {
-            // NOP to read lastAccess to update local thread values
-            if (lastAccessNanos < 0) {
-                assert true;
-            }
+            // read durationMs first to enforce a read barrier
+            int tmpDurationMs = durationMs;
 
-            return "Invocation{" + "id=" + id + ", startedAt=" + startedAt + ", durationNanos=" + durationNanos + '}';
+            return "Invocation{" + "id=" + id + ", startedAt=" + startedAt + ", durationMs=" + tmpDurationMs + '}';
         }
 
         @Override
         public JsonObject toJson() {
-            // NOP to read lastAccess to update local thread values
-            if (lastAccessNanos < 0) {
-                assert true;
-            }
+            // read durationMs first to enforce a read barrier
+            int tmpDurationMs = durationMs;
```

Comment:
`id` and `startedAt` are written after the construction of `Invocation` (to could keep the constructor function constant). So they may not be visible for the reader thread without synchronization. This should be done by the volatile `durationMs`.

So I just keep it by the book and always write `durationMs` last and read it first. Maybe this can be optimized further, since `id` and `startedAt` never change, but I'd like to stick with the patterns you can find in the books.